### PR TITLE
CSV Generator Tool

### PR DESCRIPTION
**Pull request by davidvossel**
**What this PR does / why we need it**:

Every project that integrates with the HCO needs a csv-generator binary/script shipped with their operator container image. This script is used by the HCO to generate the corresponding CSV information (which contains info on how to deploy cdi-operator, the necessary rbac, and crd ownership info).

## Changes
* Manifest templator now generates the CSV from source rather than manually generated CSV yaml file. 
* csv-generator tool (which uses the CSV generation logic) is added to cdi-operator container image. 

**Release note**:
```release-note
CSV generator tool (usr/bin/csv-generator) included in cdi-operator container image. 
```

### CODE REVIEW

**Comment by davidvossel** [comment]
fixed 'make test-lint' related errors that travis-ci was complaining about

**Comment by awels** [comment]
@davidvossel thanks for the PR, once I get some time to breathe, I will review.

File: hack/build/build-go.sh [hunk 1]
```diff
@@ -62,6 +62,11 @@ elif [ "${go_opt}" == "build" ]; then
             ln -sf ${outFile} ${outLink}
         )
     done
+
+    # This container also needs the csv-generator file, so it needs to be
+    # copied into the operator's directory for docker build to source
+    # csv-generator correctly
+    cp ${OUT_DIR}/tools/${CSV_GENERATOR}/${CSV_GENERATOR} ${OUT_DIR}/cmd/${OPERATOR}/${CSV_GENERATOR}
```
Comment (mhenriks) [inline hack/build/build-go.sh]: I wonder if this would be better handled elsewhere?  I don't think it's a big deal because we typically build everything but technically `${OUT_DIR}/tools/${CSV_GENERATOR}/${CSV_GENERATOR}` is not always guaranteed to exist.

Comment (davidvossel) [reply hack/build/build-go.sh]: yeah, this is an unfortunate hack. I couldn't figure out another place to put it... I could add a check to only copy it when OUT_DIR/tools/CSV_GENERATOR/CSV_GENERATOR exists to make it safer though. Would that make sense?

Comment (annastopel) [reply hack/build/build-go.sh]: It is possible add this copy to ``` ./hack/build/build-copy-artifacts.sh``` as it is done for cloner, or as another option (I prefer this one ) to add a dedicated script ```hack/build/build-cdi-operator.sh ``` that should be added to build target in Makefile and will copy all the required artifacts to proper location

Comment (annastopel) [reply hack/build/build-go.sh]: General question - how HCO will execute csv-generator binary? will it exec -ti to operator container ?

Comment (davidvossel) [reply hack/build/build-go.sh]: This binary will be executed during the HCO's build process when the unified CSV is generated. so, something like `docker run -it -rm --entrypoint=/usr/bin/csv-generator <imagename> <args>`

The issue is the HCO doesn't currently have the ability to accurately populate what cdi operators's Deployment, RBAC, CRDs need to be in the unified CSV file because these values have the potential to change between every CDI build. To fix this, we're using the build artifact  (cdi-operator container image in this case) to be the source of truth.

Basically, the HCO knows what cdi-operator container image it is integrating with, so it can then use that container image to "discover" the Deployement, RBAC rules, CRDs associated with the cdi-operator by executing the csv-generator binary/script.

File: pkg/operator/resources/operator/operator.go [hunk 2]
```diff
@@ -631,3 +636,240 @@ func createPrometheusPorts() *[]corev1.ContainerPort {
...
+				StrategySpecRaw: json.RawMessage(strategySpecJsonBytes),
+			},
+			CustomResourceDefinitions: csvv1.CustomResourceDefinitions{
+
+				Owned: []csvv1.CRDDescription{
+					{
+						Name:        "cdis.cdi..kubevirt.io",
+						Version:     "v1alpha1",
+						Kind:        "CDI",
+						DisplayName: "CDU deployment",
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/operator/resources/operator/operator.go]: fixed

File: tools/manifest-generator/manifest-generator.go [hunk 3]
```diff
@@ -95,6 +95,13 @@ func main() {
 	})
 
 	if *templFile != "" {
+		if *replacesCsvVersion == "" {
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply tools/manifest-generator/manifest-generator.go]: I shuffled the logic around a bit to try and accommodate this.  Let me know if it looks better now.

**Comment by davidvossel** [comment]
> It is possible add this copy to ./hack/build/build-copy-artifacts.sh as it is done for cloner, or as another option (I prefer this one ) to add a dedicated script hack/build/build-cdi-operator.sh that should be added to build target in Makefile and will copy all the required artifacts to proper location

I put the logic into the `hack/build/build-copy-artifacts.sh` file next to the similar logic that exists for cloner.

File: pkg/operator/resources/operator/operator.go [hunk 4]
```diff
@@ -631,3 +636,240 @@ func createPrometheusPorts() *[]corev1.ContainerPort {
...
+							},
+							{
+								Description:  "The targeted version of the CDI deployment.",
+								DisplayName:  "Target CDI Version",
+								Path:         "targetVersion",
+								XDescriptors: []string{"urn:alm:descriptor:text"},
+							},
+							{
+								Description:  "The observed registry of the KubeVirt deployment.",
+								DisplayName:  "Observed KubeVirt registry",
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/operator/resources/operator/operator.go]: I removed this entry.

This was a copy paste error. I had used some KubeVirt code as a primer for these cdi changes and missed removing that one line. 

For CDI, i'm not sure how important this field really is. I added it to KubeVirt for completion's sake, but realistically I can't think of many use cases outside of development where tracking the repo is all that important.

**Comment by davidvossel** [comment]
@annastopel When you have a moment, can you do another review round on this?

fyi, it's not super urgent, but we are working against some time constraints here. The introduction of all these csv-generator scripts needs to be finalized soon in order for the HCO to begin the integration work.
